tests/simulation/sim_test.go

```diff
@@ -251,3 +277,301 @@ func BenchmarkFullAppSimulation(b *testing.B) {
+	config, db, dir, _, skip, err := simapp.SetupSimulation("leveldb-app-sim", "Simulation")
+	if skip {
+		t.Skip("skipping application simulation")
+	}
+
+	require.NoError(t, err, "simulation setup failed")
+
+	defer func() {
+		db.Close()
+		require.NoError(t, os.RemoveAll(dir))
+	}()
+
+	var logger log.Logger
+	if simapp.FlagVerboseValue {
+		logger = server.ZeroLogWrapper{
+			Logger: zerolog.New(os.Stderr).Level(zerolog.InfoLevel).With().Timestamp().Logger(),
+		}
+	} else {
+		logger = server.ZeroLogWrapper{
+			Logger: zerolog.Nop(),
+		}
+	}
+
+	app := umeeapp.New(
+		logger,
+		db,
+		nil,
+		true,
+		map[int64]bool{},
+		dir,
+		simapp.FlagPeriodValue,
+		umeeapp.MakeEncodingConfig(),
+		umeeapp.EmptyAppOptions{},
+		umeeapp.GetWasmEnabledProposals(),
+		umeeapp.EmptyWasmOpts,
+		fauxMerkleModeOpt,
+	)
+	require.Equal(t, appparams.Name, app.Name())
+
+	// run randomized simulation
+	stopEarly, simParams, simErr := simulation.SimulateFromSeed(
+		t,
+		os.Stdout,
+		app.BaseApp,
+		appStateFn(app.AppCodec(), app.StateSimulationManager),
+		simtypes.RandomAccounts,
+		simapp.SimulationOperations(app, app.AppCodec(), config),
+		app.ModuleAccountAddrs(),
+		config,
+		app.AppCodec(),
+	)
+
+	// export state and simParams before the simulation error is checked
+	err = simapp.CheckExportSimulation(app, config, simParams)
+	require.NoError(t, err)
+	require.NoError(t, simErr)
+
+	if config.Commit {
+		simapp.PrintStats(db)
+	}
+
+	if stopEarly {
+		fmt.Println("can't export or import a zero-validator genesis, exiting test...")
+		return
+	}
+
+	fmt.Printf("exporting genesis...\n")
+
+	exported, err := app.ExportAppStateAndValidators(false, []string{})
+	require.NoError(t, err)
+
+	fmt.Printf("importing genesis...\n")
+
+	config, newDB, newDir, _, _, err := simapp.SetupSimulation("leveldb-app-sim-2", "Simulation-2")
+	require.NoError(t, err, "simulation setup failed")
+
+	defer func() {
+		require.NoError(t, newDB.Close())
+		require.NoError(t, os.RemoveAll(newDir))
+	}()
+
+	newApp := umeeapp.New(
+		logger,
+		newDB,
+		nil,
+		true,
+		map[int64]bool{},
+		newDir,
+		simapp.FlagPeriodValue,
+		umeeapp.MakeEncodingConfig(),
+		umeeapp.EmptyAppOptions{},
+		umeeapp.GetWasmEnabledProposals(),
+		umeeapp.EmptyWasmOpts,
+		fauxMerkleModeOpt,
+	)
+	require.Equal(t, appparams.Name, newApp.Name())
+
+	var genesisState GenesisState
+	err = json.Unmarshal(exported.AppState, &genesisState)
+	require.NoError(t, err)
+
+	// newApp.InitChain(abci.RequestInitChain{
+	// 	AppStateBytes: exported.AppState,
+	// })
+
+	defer func() {
+		if r := recover(); r != nil {
+			err := fmt.Sprintf("%v", r)
+			if !strings.Contains(err, "1") {
+				panic(r)
+			}
+			logger.Info("Skipping simulation as all validators have been unbonded")
+			logger.Info("err", err, "stacktrace", string(debug.Stack()))
+		}
+	}()
+
+	ctxB := newApp.NewContext(true, tmproto.Header{Height: app.LastBlockHeight()})
+	newApp.ModuleManager.InitGenesis(ctxB, newApp.AppCodec(), genesisState)
```

Comment:
Instead of exposing `ModuleManager`, how about exposing `InitGenesis` function? Or, ideally, init genesis through `app.InitChainer` - you just need to provide `genesisState` into the request.

Reply:
OK